index.d.ts

```diff
@@ -0,0 +1,486 @@
+
+    /**
+    * Session
+    * @see { @link https://realm.io/docs/javascript/latest/api/Realm.Sync.Session.html }
+    */
+    interface Session {
+        readonly config: any;
+        readonly state: string;
+        readonly url: string;
+        readonly user: User;
+    }
+
+    /**
+    * AuthError
+    * @see { @link https://realm.io/docs/javascript/latest/api/Realm.Sync.AuthError.html }
+    */
+    interface AuthError {
+        readonly code: number;
+        readonly type: string;
+    }
+
+    /**
+     * ChangeEvent
+     * @see { @link https://realm.io/docs/javascript/latest/api/Realm.Sync.ChangeEvent.html }
+     */
+    interface ChangeEvent {
+        readonly changes: { [object_type: string]: { insertions: number[], deletions: number[], modifications: number[] } };
+        readonly oldRealm: Realm;
+        readonly path: string;
+        readonly realm: Realm;
+    }
+
+    function addListener(serverURL: string, adminUser: Realm.Sync.User, regex: string, name: string, changeCallback: () => void): void;
+    function removeAllListeners(name?: string[]): void;
+    function removeListener(regex: string, name: string, changeCallback: () => void): void;
+    function setLogLevel(logLevel: LogLevelType): void;
+    function setAccessToken(accessToken: string): void;
+
+    type Instruction = {
+        type: 'INSERT' | 'SET' | 'DELETE' | 'CLEAR' | 'LIST_SET' | 'LIST_INSERT' | 'LIST_ERASE' | 'LIST_CLEAR' | 'ADD_TYPE' | 'ADD_PROPERTIES'
+        object_type: string,
+        identity: string,
+        values: any | undefined
+        list_index: any | undefined
+        object_identity: any | undefined
+        new_identity: any | undefined,
+        property: any | undefined,
+        properties: any | undefined,
+        primary_key: string | undefined
+    }
+
+    class Adapter {
+        constructor(
+            local_path: string,
+            server_url: string,
+            admin_user: User,
+            regex: string,
+            change_callback: Function
+        )
+
+        /**
+         * Advance the to the next transaction indicating that you are done processing the current instructions for the given Realm.
+         * @param path the path for the Realm to advance
+         */
+        advance(path: string): void;
+        close(): void;
+        current(path: string): Array<Instruction>;
+        realmAtPath(path: string, realmID?: string, schema?: ObjectSchema[]): Realm
+    }
+}
+
+declare class Realm {
+    static defaultPath: string;
+
+    readonly path: string;
+    readonly readOnly: boolean;
+    readonly schema: Realm.ObjectSchema[];
+    readonly schemaVersion: number;
+
+    syncSession: Realm.Sync.Session;
+
+    /**
+     * @param  {string} path
+     * @param  {any} encryptionKey?
+     * @returns number
+     */
+    static schemaVersion(path: string, encryptionKey?: any): number;
+
+    /**
+     * @param  {Realm.Configuration} config?
+     */
+    constructor(config?: Realm.Configuration);
+
```

Comment:
The alternative `open` and `openAsync` factories seem to be missing.
https://github.com/realm/realm-js/blob/master/lib/extensions.js#L45